src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -466,59 +444,115 @@ def result_to_dict(value: Any) -> Any:
+    if SiloMode.get_current_mode() == SiloMode.MONOLITH:
+        # Pass the response through json.loads and json.dumps to force testing the serialization of services
+        # in monolith mode -- if this throws an exception, it means the types of an rpc service have an issue.
+        # do not remove!
+        serial_response = json.loads(
+            json.dumps(dispatch_to_local_service(service_name, method_name, serial_arguments))
+        )
+    else:
+        serial_response = _dispatch_to_silo_service(
+            method_name, region, serial_arguments, service_name, use_test_client
+        )
 
+    return_value = serial_response["value"]
+    return (
+        None
+        if return_value is None
+        else service.deserialize_rpc_response(method_name, return_value)
+    )
+
+
+def _dispatch_to_silo_service(method_name, region, serial_arguments, service_name, use_test_client):
     if region is None:
         address = settings.SENTRY_CONTROL_ADDRESS
     else:
         address = region.address
-
     if not (address and settings.RPC_SHARED_SECRET):
         raise RpcSendException("Not configured for RPC network requests")
-
     path = django.urls.reverse(
         "sentry-api-0-rpc-service",
         kwargs={"service_name": service_name, "method_name": method_name},
     )
     url = address + path
-
     request_body = {
         "meta": {},  # reserved for future use
         "args": serial_arguments,
     }
-
+    data = json.dumps(request_body).encode(_RPC_CONTENT_CHARSET)
+    signature = generate_request_signature(path, data)
+    headers = {
+        "Content-Type": f"application/json; charset={_RPC_CONTENT_CHARSET}",
+        "Authorization": f"Rpcsignature {signature}",
+    }
     timer = metrics.timer(
         "hybrid_cloud.dispatch_rpc.duration", tags={"service": service_name, "method": method_name}
     )
     span = sentry_sdk.start_span(
         op="hybrid_cloud.dispatch_rpc", description=f"rpc to {service_name}.{method_name}"
     )
     with span, timer:
-        response = _fire_request(url, path, request_body)
+        if use_test_client:
+            response = _fire_test_request(path, headers, data, region)
+        else:
+            response = _fire_request(url, headers, data)
         metrics.incr(
             "hybrid_cloud.dispatch_rpc.response_code", tags={"status": response.status_code}
         )
 
-    serial_response = response.json()
-    return_value = serial_response["value"]
-    return (
-        None
-        if return_value is None
-        else service.deserialize_rpc_response(method_name, return_value)
+        if response.status_code != 200:
+            if in_test_environment():
+                if response.status_code == 500:
+                    raise Exception(
+                        "Error invoking rpc at {path}: check error logs for more details"
+                    )
+                raise Exception("Error invoking rpc at {path}: {response.json()['detail']}")
```

Comment:
The `f` is missing from the beginning of the strings.

Also, nit: adding an `!r` to the interpolations would make them a bit more readable (as in `{path!r}`, which is equivalent to `{repr(path)}`).